blssig/verifier.go

```diff
@@ -103,5 +106,43 @@ func (v *Verifier) Verify(pubKey gpbft.PubKey, msg, sig []byte) (_err error) {
 		return fmt.Errorf("unarshalling public key: %w", err)
 	}
 
-	return v.scheme.Verify(point, msg, sig)
+	return v.bdnScheme.Verify(point, msg, sig)
+}
+
+func (v *Verifier) BatchVerify(pubKeys []gpbft.PubKey, msgs [][]byte, sigs [][]byte) (_err error) {
+	defer func() {
+		status := measurements.AttrStatusSuccess
+		if _err != nil {
+			status = measurements.AttrStatusError
+		}
+		if perr := recover(); perr != nil {
+			_err = fmt.Errorf("panicked validating batch signature: %v\n%s",
+				perr, string(debug.Stack()))
+			log.Error(_err)
+			status = measurements.AttrStatusPanic
+		}
+		metrics.verify.Add(context.TODO(), 1, metric.WithAttributes(status))
+	}()
+	if len(msgs) != len(sigs) {
+		return fmt.Errorf("number of signatures must match the number of messages")
+	}
+
+	if len(pubKeys) != len(sigs) {
+		return fmt.Errorf("number of public keys must match the number of messages")
+	}
+
+	points := make([]kyber.Point, len(msgs))
+	for i, pubKey := range pubKeys {
+		var err error
+		points[i], err = v.pubkeyToPoint(pubKey)
+		if err != nil {
+			return fmt.Errorf("unarshalling public key: %w", err)
+		}
+	}
+	aggSig, err := v.blsScheme.AggregateSignatures(sigs...)
```

Comment:
I don't think this is generally safe. It might be secure in this case as the message and justification are guaranteed to be unique.
No, wait, they are not. Then, yeah, this is not safe. The BatchVerify method requires that the messages be unique. There is an alternative method for BatchVerifying that introduces randomness to both the message and signature, making them linearly independent. I could write that.
It does not affect the overall scheme and is just a verification time tweak.

Reply:
Hmm, I guess the BatchVerify will check that messages are unique. So it might be safe.